cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
+          size += rawMessage.getId().length() + rawMessage.getPayload().length + 8;
+        }
+
+        encoder.setItemCount(monitorMessages.size());
+        for (GenericRecord monitorMessage : monitorMessages) {
+          encoder.startItem();
+          messageWriter.write(monitorMessage, encoder);
+        }
+
+        if (!iterator.hasNext()) {
+          // close this iterator before moving to next one.
+          iterator.close();
+          encoder.writeArrayEnd();
         }
       }
 
-      jsonWriter.endObject();
-    }
+      // end the map when all the request entries have been
+      if (!requestsIterator.hasNext()) {
+        encoder.writeMapEnd();
+        mapEnded = true;
+      }
 
-    if (buffer.isReadable()) {
-      chunkResponder.sendChunk(buffer.copy());
+      return chunk.copy();
     }
 
-    Closeables.closeQuietly(chunkResponder);
-  }
+    private CloseableIterator<Message> fetchMessages(Map.Entry<String, GenericRecord> requestEntry)
+      throws TopicNotFoundException, IOException {
+      String topic = requestKeyToLocalTopic.computeIfAbsent(requestEntry.getKey(), this::getTopic);
 
-  private String getTopic(String topicConfig) {
-    int idx = topicConfig.lastIndexOf(':');
-    return idx < 0 ? cConf.get(topicConfig) : cConf.get(topicConfig.substring(0, idx)) + topicConfig.substring(idx + 1);
-  }
+      int limit = (int) requestEntry.getValue().get("limit");
+      String fromMessage = requestEntry.getValue().get("messageId") == null ? null :
+        requestEntry.getValue().get("messageId").toString();
+      return messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(), topic, limit, fromMessage);
+    }
 
-  @POST
-  @Path("/shutdown")
-  public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
-    responder.sendString(HttpResponseStatus.OK, "Triggering shutdown down Runtime Http Server.");
-    shutdownRunnable.run();
-  }
+    @Override
+    public void finished() throws Exception {
+      chunk.release();
+    }
 
-  private void fetchAndWriteMessages(JsonWriter jsonWriter, ByteBuf buffer,
-                                     ChunkResponder chunkResponder, String topic, int limit,
-                                     @Nullable String fromMessage) throws TopicNotFoundException, IOException {
-    try (CloseableIterator<Message> iter = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(), topic, limit,
-                                                                fromMessage)) {
-      while (iter.hasNext()) {
-        Message message = iter.next();
-        GSON.toJson(new MonitorMessage(message.getId(), message.getPayloadAsString(StandardCharsets.UTF_8)),
-                    MonitorMessage.class, jsonWriter);
-        if (buffer.readableBytes() >= CHUNK_SIZE) {
-          chunkResponder.sendChunk(buffer.copy());
-          buffer.clear();
-        }
+    @Override
+    public void handleError(@Nullable Throwable cause) {
+      while (iterator.hasNext()) {
```

Comment:
Why it is a `while`?? I thought you can always just close the current `iterator`??